wandb/sdk/launch/runner/kubernetes_runner.py

```diff
@@ -406,3 +696,63 @@ def maybe_create_imagepull_secret(
         return core_api.create_namespaced_secret(namespace, secret)
     except Exception as e:
         raise LaunchError(f"Exception when creating Kubernetes secret: {str(e)}\n")
+
+
+def add_wandb_env(root: Union[dict, list], env_vars: Dict[str, str]) -> None:
```

Comment:
I think it'd be safer to make sure this is modifying the right "containers" key. As it is, this could have weird effects if users have an [annotation](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) called "containers" in their spec.

Recommendations:
- Rename `root` -> `resource_args`
- Do something like:
```
containers = resource_args.get("spec", {}).get("template", {}).get("spec", {}).get("containers", [])
for container in containers:
  <inject env vars>
```

Reply:
And similar with `add_label_to_pods`

Reply:
The problem is that this needs to be able to handle custom resources that don't necessarily follow that structure. For example volcano jobs put container definitions under spec.tasks.template.spec.container, and kubeflow does spec.pyTorchReplicaSpecs.{Master,Worker}.template.spec.containers

My hope is that this is okay as long we document the behavior. It should only be if you have "containers" key under a "spec" key and I feel like that criteria won't have a lot of false positives

Reply:
Agreed about renaming root to resource args though

Reply:
Ah that makes sense. In that case I'm ok leaving this as it is. You could also consider explicitly allowing the three formats ("standard" k8s, kubeflow, and volcano), but the recursive solution is way simpler.